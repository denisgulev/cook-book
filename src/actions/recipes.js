import uuid from "uuid";
import db from "../firebase/firebase";

/* WHITOUT DB */
/**
 * component calls action generator
 * action generator returns object
 * component dispatches object
 * redux store changes
 */

/* WITH DB */
/**
 * components calls action generator
 * action generator returns function
 * component dispatches function (?)
 * function runs (has the ability to dispatch other actions)
 */

// add recipe

export const addRecipe = recipe => ({
  type: "ADD_RECIPE",
  recipe
});

export const startAddRecipe = (recipeData = {}) => {
  return dispatch => {
    const {
      title = "",
      description = "",
      note = "",
      createdAt = 1000
    } = recipeData;

    const recipe = { title, description, note, createdAt };

    return db
      .ref("recipes")
      .push(recipe)
      .then(ref => {
        dispatch(
          addRecipe({
            id: ref.key,
            ...recipe
          })
        );
      });
  };
};

// remove recipe

export const removeRecipe = ({ id } = {}) => ({
  type: "REMOVE_RECIPE",
  id
});

// edit recipe

export const editRecipe = (id, updates) => ({
  type: "EDIT_RECIPE",
  id,
  updates
});
