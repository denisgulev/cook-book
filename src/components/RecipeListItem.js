import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

const RecipeListItem = ({ id, title, description, createdAt }) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{title}</h3>
      <span className="list-item__subtitle">{description}</span>
    </div>
    <h3 className="list-item__data">
      {moment(createdAt).format("MMMM Do, YYYY")}
    </h3>
  </Link>
);

export default RecipeListItem;
