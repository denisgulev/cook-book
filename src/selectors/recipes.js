import moment from "moment";

// get visibleRecipes - 2nd argument is "filterReducerState"
const getVisibleRecipes = (recipes, { text, sortBy, startDate, endDate, category }) => {
  return recipes
    .filter(recipe => {
      const textMatch = recipe.title.toLowerCase().includes(text.toLowerCase());
      if (sortBy === "date") {
        const createdAtMoment = moment(recipe.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;

        return startDateMatch && endDateMatch && textMatch;
      } else if (sortBy === "category") {
        if (recipe.category !== undefined) {
          if (category.toLowerCase() === "all") {
            // show all recipes
            return true && textMatch;
          } else {
            // show recipe with category = searched category
            const categoryMatch = recipe.category.toLowerCase() === category.toLowerCase();

            return categoryMatch && textMatch;
          }
        } else {
          return false;
        }
      }
    })
    .sort((a, b) => {
      if (sortBy === "date") return a.createdAt < b.createdAt ? 1 : -1;
      else if (sortBy === "category") return a.title < b.title ? -1 : 1;
    });
};

export default getVisibleRecipes;
