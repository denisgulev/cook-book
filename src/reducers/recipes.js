import * as actions from "../actions/actionTypes";

const recipesReducerDefaultState = [];

const recipesReducer = (state = recipesReducerDefaultState, action) => {
    switch (action.type) {
        case actions.ADD_RECIPE:
            return [...state, action.recipe];
        case actions.REMOVE_RECIPE:
            return state.filter(({id}) => id !== action.id);
        case actions.EDIT_RECIPE:
            return state.map(recipe => {
                if (recipe.id === action.id) {
                    return {
                        ...recipe,
                        ...action.updates
                    };
                } else {
                    return recipe;
                }
            });
        case actions.SET_RECIPES:
            return action.recipes;
        default:
            return state;
    }
};

export default recipesReducer;
