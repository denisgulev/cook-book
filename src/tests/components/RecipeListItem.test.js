import React from "react";
import { shallow } from "enzyme";
import recipes from "../fixtures/recipes";
import RecipeListItem from "../../components/RecipeListItem";

test("should render RecipeListItem correctly", () => {
  const wrapper = shallow(
    <RecipeListItem key={recipes[0].id} {...recipes[0]} />
  );
  expect(wrapper).toMatchSnapshot();
});
