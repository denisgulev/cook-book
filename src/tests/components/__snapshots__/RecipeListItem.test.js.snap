// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RecipeListItem correctly 1`] = `
<Link
  className="list-item"
  to="/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Ricetta in selector test
    </h3>
    <span
      className="list-item__subtitle"
    >
      How to do in selector test
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    January 1st, 1970
  </h3>
</Link>
`;
